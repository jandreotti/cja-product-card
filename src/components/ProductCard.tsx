import styles from '../styles/styles.module.css';

import { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  InitialValuesInterface,
  onChangeArgs,
  Product,
  ProductContextProps,
  ProductCardHandlers,
} from '../interfaces/ProductInterfaces';
import React from 'react';

// CONTEXTO -> se usa para pasar datos entre componentes (Puedo pasarle del padre a los hijos)
// Lo dejo en este archivo porque solamente voy a usar el contexto en el ProductCard
export const ProductContext = createContext({} as ProductContextProps); // la informacion que le voy a compartir a tdoos los hijos
const { Provider } = ProductContext;

// INTERFACE DE LAS PROPS de ProductCard
export interface Props {
  product: Product;
  //children?: ReactElement | ReactElement[]; // Aca se pone que el componente va a contener hijos
  children: (args: ProductCardHandlers) => JSX.Element; //! Esto sirve para poder tener un hijo que sea una funcion que devuelva un JSX.Element. los args son los valores del hijo que estoy exponiendo al padre
  //
  className?: string;
  style?: CSSProperties;
  //
  onChange?: (args: onChangeArgs) => void; // FUNCION MARAVILLA //Asi es la funcion que yo estoy esperando
  value?: number; // VALOR DE LA CANTIDAD DE PRODUCTOS QUE TENGO EN EL CARRITO. Es opcional porque puede ser que el usuario lo especifique o no. puede ser que el usuario quiera trabajarlo en base al onChange de arriba, etc le damos la posibilidad a la persona...
  //
  initialValues?: InitialValuesInterface; //! VALORES INICIALES
}
export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  // CUSTOM HOOK
  const {
    counter,
    increasedBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    // FUNCION MARAVILLA
    onChange,
    product,
    value,
    initialValues, //! le mando el initialValues al hook quemaneja el valor inicial
  });

  // RETURN
  return (
    <div className={`${styles.productCard} ${className}`} style={style}>
      <Provider
        value={{
          // Que informacion necesitan los hijos?
          counter,
          increasedBy,
          product,
          maxCount,
        }}
      >
        {/* {()=>("123")} */}
        {/* {children} */}
        {/* //! Debo cambiar esto para que funcione con los hijos que sean funciones. Porque no puedo reenderizar funciones. lo que puedo hacer es ejecutar funciones */}
        {/* Como esto es una funcion que puedo ejecutar, puedo mandar cualquier cantodad de argumentos y estos argumentos van a la capa superior o PADRE. */}
        {/* OSea que le mando la informacion ya procesada y masticada a mi componente  padre */}
        {/* //! Aca debo mandar todo lo que tengo que exponer al exterior */}
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          increasedBy,
          reset,
        })}
      </Provider>
    </div>
  );
};
