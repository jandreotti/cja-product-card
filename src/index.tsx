//! EXPORTAR TODO lo de la carpeta components
export * from './components';
