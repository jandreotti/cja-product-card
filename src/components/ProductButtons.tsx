import React, { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface Props {
	className?: string;
	style?: CSSProperties;
}
export const ProductButtons = ({ className = '', style }: Props) => {
	const { counter, increasedBy, maxCount } = useContext(ProductContext);

	// TODO isMaxReached: useCallback, [count, maxCount]
	//TRUE si count === maxCount
	//FALSE si no lo es

	const isMaxReached = useCallback(() => {
		//! MIO
		// if (counter === maxCount) {
		// 	return true;
		// }
		// return false;

		//! PROFESOR
		return !!maxCount && counter === maxCount;
	}, [counter, maxCount]);

	return (
		<div className={`${styles.buttonsContainer} ${className}`} style={style}>
			<button className={styles.buttonMinus} onClick={() => increasedBy(-1)}>
				-
			</button>

			<div className={styles.countLabel}>{counter}</div>

			<button
				// className={isMaxReached() ? `${styles.buttonAdd} ${styles.disabled}` : `${styles.buttonAdd}`}
				className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
				onClick={() => increasedBy(1)}
			>
				+
			</button>
		</div>
	);
};
