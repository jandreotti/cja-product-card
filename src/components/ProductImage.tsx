import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

//Si me manda una imagen img al componente significa que esa imagen debe sobrescribir la imagen de producto

export interface Props {
  img?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ img = '', className, style }: Props) => {
  const { product } = useContext(ProductContext);

  let imgToShow: string;
  if (img) {
    imgToShow = img;
  } else if (product.img) {
    imgToShow = product.img;
  } else {
    imgToShow = noImage;
  }

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt="Product Image"
      style={style}
    />
  );
};
