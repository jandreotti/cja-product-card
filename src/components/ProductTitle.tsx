import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface Props {
  title?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ title, className, style }: Props) => {
  const { product } = useContext(ProductContext);
  // Si viene el titulo usa el titulo, sino usa el product.title
  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {title || product.title}
    </span>
  );
};
