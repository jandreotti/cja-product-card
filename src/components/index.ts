// importo los componentes hijos
import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';

///! 2ª
// importo el componente padre
import { ProductCard as ProductCardHOC } from './ProductCard'; // importo un componente ProductCard con el nombre ProductCardHOC
import { ProductCardHOCProps } from '../interfaces/ProductInterfaces';

//! 1ª
export { ProductButtons } from './ProductButtons';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';

//! 2ª
// exporto el componente padre ProductCardHOC agregandole los componentes hijos renombrandolos para poder hacer el segundo tipo de uso del componente
export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
	Title: ProductTitle,
	Image: ProductImage,
	Buttons: ProductButtons,
});

// export default ProductCard; //Puedo ponerlo o no
