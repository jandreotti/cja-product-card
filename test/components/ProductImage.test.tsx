import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard, ProductImage } from '../../src/components';
import { product2 } from '../data/products';

describe('PRUEBAS DEL ProductImage', () => {
  //! Prueba para hacerle test a un snapshot simple del componente ProductImage
  test('debe mostrar el componente correctamente con el titulo personalizado', () => {
    const wrapper = renderer.create(<ProductImage img="imagen.jpg" />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  //! Prueba para hacerle test a un snapshot de la relacion entre ProductCard y ProductTitle
  test('debe de mostrar el componente con la imagen del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
