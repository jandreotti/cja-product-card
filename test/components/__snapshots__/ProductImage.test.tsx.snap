// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PRUEBAS DEL ProductImage debe de mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product Image"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`PRUEBAS DEL ProductImage debe mostrar el componente correctamente con el titulo personalizado 1`] = `
<img
  alt="Product Image"
  className="productImg undefined"
  src="imagen.jpg"
/>
`;
